state/prunning_service.go

```diff
@@ -0,0 +1,228 @@
+		return retainHeightInfo.Height, err
+	}
+
+	return 0, nil
+}
+func (p *Pruner) pruningRoutine() {
+	lastHeightPruned := int64(0)
+	lastABCIResPrunedHeight := int64(0)
+	for {
+		select {
+		case <-p.pruningTicker.C:
+			retainHeight := p.findMinRetainHeight()
+			if retainHeight != lastHeightPruned {
+				pruned, evRetainHeight, err := p.pruneBlocks(retainHeight)
+				p.logger.Info(fmt.Sprintf("Pruned %d, evidence retain height is %d , err %s", pruned, evRetainHeight, err))
+				lastHeightPruned = retainHeight
+			}
+
+			ABCIResRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+			if err == nil {
+				if lastABCIResPrunedHeight != ABCIResRetainHeight {
+					pruned, _ := p.stateStore.PruneABCIResponses(ABCIResRetainHeight)
+					p.logger.Info(fmt.Sprintf("Pruned %d abci responses ", pruned))
+				}
+			}
+		case <-p.Quit():
+			return
+		}
+
+	}
+}
+
+// If no retain height has been set by the application or the data companion
+// the database will not have values for the corresponding keys.
+// If both retain heights were set, we pick the smaller one
+// If only one is set we return that one
+func (p *Pruner) findMinRetainHeight() int64 {
+	var noAppRetainHeightSet, noDCRetainHeight bool
+	appRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
+
+	if err != nil {
+		if err == ErrKeyNotFound {
+			noAppRetainHeightSet = true
+		} else {
+			return 0
+		}
+	}
+	dcRetainHeight, err := p.stateStore.GetDataCompanionRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			noDCRetainHeight = true
+		} else {
+			return 0
+		}
+	}
+	if !noAppRetainHeightSet && !noDCRetainHeight {
+		if appRetainHeight < dcRetainHeight {
+			return appRetainHeight
+		}
+		return dcRetainHeight
+	}
+
+	if !noAppRetainHeightSet {
+		return appRetainHeight
+	}
+	if !noAppRetainHeightSet {
+		return dcRetainHeight
+	}
+
+	return 0
+}
+
+func (p *Pruner) pruneBlocks(height int64) (pruned uint64, evRetainHeight int64, err error) {
+	if height <= 0 {
+		return 0, 0, errors.New("retain height cannot be less or equal than 0")
+	}
+
+	base := p.bs.Base()
+
+	var state State
+	state, err = p.stateStore.Load()
+	if err != nil {
+		p.logger.Error("failed to load state, cannot prune")
+		return
+	}
+	p.logger.Info(fmt.Sprintf("Received pruning request for %d. Accepted retain height is : %d", height, height))
+	pruned, evRetainHeight, err = p.bs.PruneBlocks(height, state)
+	if err != nil {
+		p.logger.Error(fmt.Sprintf("failed to prune blocks for retain height %d with error: %s", height, err))
+	} else {
+		p.logger.Info("pruned blocks", "pruned", pruned, "retain_height", height)
+
+	}
+	err = p.stateStore.PruneStates(base, height, evRetainHeight)
+	if err != nil {
+		p.logger.Error(fmt.Sprintf("failed to prune the state store with error: %s ", err))
+	}
+	return pruned, evRetainHeight, err
+}
+
+func (p *Pruner) PruneABCIResponses(height int64) (pruned uint64, err error) {
+	pruned, err = p.stateStore.PruneABCIResponses(height)
+	return
+}
```

Comment:
```suggestion
func (p *Pruner) PruneABCIResponses(height int64) (uint64, error) {
	return p.stateStore.PruneABCIResponses(height)
}
```

Also, does this method really need to be public? If not I'd recommend getting rid of it altogether in favour of accessing `p.stateStore.PruneABCIResponses` directly.

Reply:
Actually this is dead code I did not delete. The pruning function is using the `stateStore.PruneABCIResponses` directly.